redis-6.0.9/src/modules/systemtime.c

```diff
@@ -0,0 +1,246 @@
+
+    // setup time and timer related stuff
+    // note: timer is _not_ armed here!
+    if (0 != setup_timer(ctx)){
+		RedisModule_Log(ctx, "error", "setting up timer related stuff failed");
+        return REDISMODULE_ERR;
+    }
+
+	if (pthread_mutex_init(&lock, NULL) != 0) {
+		RedisModule_Log(ctx, "error", "pthread_mutex_init failed");
+        return REDISMODULE_ERR; 
+    }
+
+    // start the collector thread
+    if (0 > start_sys_time_collector_thread(ctx)){
+		RedisModule_Log(ctx, "error", "starting the collector thread failed");
+        return REDISMODULE_ERR;
+    }
+
+    if (RedisModule_Init(ctx,"drivenetsgettime",1,REDISMODULE_APIVER_1) == REDISMODULE_ERR){
+		RedisModule_Log(ctx, "error", "RedisModule_Init failed");
+		return REDISMODULE_ERR;
+	}
+
+	// registering with redis!!
+	if (RedisModule_CreateCommand(ctx,"drivenets.get.time",
+                                  DrivenetsGetTime_RedisCommand,
+								  "readonly",0,0,0)              == REDISMODULE_ERR){
+
+		RedisModule_Log(ctx, "error", "DrivenetsGetTime_RedisCommand failed");
+        return REDISMODULE_ERR;
+	}
+
+	return REDISMODULE_OK;
+}
+
+int setup_io_multiplexer(RedisModuleCtx *ctx)
+{
+    // we want a multiplexer of size one.
+    // just one timeout to manage
+    if (0 > (s_epoll_fd = epoll_create(1))){
+		RedisModule_Log(ctx, "error", "epoll_create failed. errno: %d", errno);
+		return -1;
+	}
+ 
+	/* create timer */
+    // CLOCK_MONOTONIC is a good clock to use here, as its
+    // man states that it is _not_ sensetive to changes
+    // we want a non-blocking fd, so that we can avoid blocking
+    // on the read when we leave the epoll on a _non_ timer related event
+	s_timer_fd = timerfd_create(CLOCK_MONOTONIC, TFD_CLOEXEC | TFD_NONBLOCK);
+	if (0 > s_timer_fd){
+		RedisModule_Log(ctx, "error", "timerfd_create failed. errno: %d", errno);
+		return -1;
+	}
+	
+    // add the timer to the multiplexer
+    struct epoll_event ev;
+	ev.events = EPOLLIN;
+	ev.data.fd = s_timer_fd;
+	int sts = epoll_ctl(s_epoll_fd, EPOLL_CTL_ADD, s_timer_fd, &ev);
+	if (0 > sts){
+		RedisModule_Log(ctx, "error", "epoll_ctl failed. errno: %d", errno);
+		return -1;
+	}
+	return 0;
+}
+
+int setup_timer(RedisModuleCtx *ctx)
+{
+    struct itimerspec its;
+	// according to time(7) the accuracy of systemcalls that set timers
+	// like sleep(2) is given by a unit called jiffi. the size of
+	// one jiffi is equal to a kernel constat that indicates number of
+	// clock ticks per second.
+	// take this constant from the system as it is coupled to the os / hw versions
+	long hz = sysconf(_SC_CLK_TCK);
+	if (0 > hz){
+		RedisModule_Log(ctx, "error", "sysconf failed. errno: %d", errno);
+		return -1;
+	}
+	// one jiffi in seconds, is therefor:
+	jiffi_in_seconds = 1.0 / hz;
+
+	//printf("jiffi_in_seconds: %f\n", jiffi_in_seconds);
+ 
+	its.it_interval.tv_sec = 0;
+	its.it_interval.tv_nsec = jiffi_in_seconds * NANO_IN_SEC;
+	its.it_value.tv_sec = 1; // we start aligned to a full second
```

Comment:
the request was to show time with minute accuracy, so you doing a lot of not needed work here

Reply:
yeah ... that was a bit of an overkill :)